dev/tests/static/testsuite/Magento/Test/Legacy/Magento/Framework/App/Action/AbstractActionTest.php

```diff
@@ -0,0 +1,139 @@
+    }
+
+    /**
+     * Test newly created controllers do not extend deprecated AbstractAction.
+     *
+     * @throws \ReflectionException
+     */
+    public function testNewControllersDoNotExtendAbstractAction(): void
+    {
+        $files = $this->getTestFiles();
+
+        $found = $this->childrenClassesSearch->getClassesWhichAreChildrenOf($files, AbstractAction::class);
+
+        $this->assertEmpty(
+            $found,
+            "The following new controller(s) extend " . AbstractAction::class . "\r\n"
+            . "All new controller classes must implement " . ActionInterface::class . " instead.\r\n"
+            . print_r($found, true)
+        );
+    }
+
+    /**
+     * Provide files for test.
+     *
+     * @return array
+     */
+    private function getTestFiles(): array
+    {
+        $phpFiles = self::getAddedFilesList(self::getChangedFilesBaseDir());
+
+        $phpFiles = Files::composeDataSets($phpFiles);
+        $fileTypes = Files::INCLUDE_APP_CODE | Files::INCLUDE_LIBS | Files::AS_DATA_SET;
+        return array_intersect_key($phpFiles, $this->fileUtilities->getPhpFiles($fileTypes));
+    }
+
+    /**
+     * Provide list of new files.
+     *
+     * @param $changedFilesBaseDir
+     *
+     * @return string[]
+     */
+    private static function getAddedFilesList($changedFilesBaseDir)
+    {
+        return self::getFilesFromListFile(
+            $changedFilesBaseDir,
+            'changed_files*.added.*',
+            function () {
+                // if no list files, probably, this is the dev environment
+                // phpcs:ignore Generic.PHP.NoSilencedErrors,Magento2.Security.InsecureFunction
+                @exec('git diff --cached --name-only --diff-filter=A', $addedFiles);
+                return $addedFiles;
+            }
+        );
+    }
+
+    /**
+     * Read files from generated lists.
+     *
+     * @param string $listsBaseDir
+     * @param string $listFilePattern
+     * @param callable $noListCallback
+     * @return string[]
+     */
+    private static function getFilesFromListFile(
+        string $listsBaseDir,
+        string $listFilePattern,
+        callable $noListCallback
+    ): array {
+        $filesDefinedInList = [];
+        $listFiles = glob($listsBaseDir . '/_files/' . $listFilePattern);
+        if (!empty($listFiles)) {
+            foreach ($listFiles as $listFile) {
+                $filesDefinedInList[] = file($listFile, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
+            }
+            $filesDefinedInList = array_merge([], ...$filesDefinedInList);
+        } else {
+            $filesDefinedInList = call_user_func($noListCallback);
+        }
+        array_walk(
+            $filesDefinedInList,
+            function (&$file) {
+                $file = BP . '/' . $file;
+            }
+        );
+        $filesDefinedInList = array_values(array_unique($filesDefinedInList));
+
+        return $filesDefinedInList;
+    }
```

Comment:
I see those methods were copy-pasted from https://github.com/magento/magento2/blob/91bead52b3063c9c022d9809ee4192e1b28c7862/dev/tests/static/testsuite/Magento/Test/Php/LiveCodeTest.php.

I think it's better to extract this code to some external class and use it in both cases in order to prevent code duplication and fixing issues in one place, but keeping in other?

Reply:
Moved to separated files.
Tests have been added.